### PR TITLE
Array grammar updates

### PR DESCRIPTION
**Pull request by jakelishman**
### Summary

This is a reference implementation of arrays in the ANTLR grammar and the reference AST.

It adds an new element (`indexExpression`) into the expression tree, which is used in the general case for indexing operations.  Before I made more wide-ranging changes to `indexIdentifier`, that was the simplest way to implement multiple successive indices (and I would argue the correct way anyway), but it's also necessary to support constructs such as `bit[8](x)[0:2]`, where the index operator is applied to the result of another element of the expression tree.

There are lots of places in the current grammar where `indexIdentifier` is used, even though a simple `expression` is probably better grammatically, with the semantics left to a later point.  I haven't changed those, in the interests of keeping changes strictly to the relevant set.

This PR includes the same changes as in Qiskit/openqasm#318; I expect that to be reviewed and merged before this hits master.

It's probably easiest to review the commits separately:
- 1962f2b (simple) just adds a files to `examples` with new array things in it.
- ceeefe5 (big) the changes to the ANTLR grammar, and updates all the tests to accommodate the new expression tree (and additional grammar elements)
- a7bfcde (simple) adds the new keywords and bits to the vim plugin
- 303cad9 (big) updates the reference AST and ANTLR tree visitor to build the new elements into the AST. 

### Details and comments

(commit message)

This necessitated more wide-ranging changes than might have been
expected, because it needed to overcome various prior assumptions that
arrays would be single-dimensioned only.  There were portions where
`indexIdentifier` was defined in order to be used in `let` aliases, but
then reused in various other places.  This caused issues where things
like `a ++ b` were valid in places where it should not have been, and
regular indexing (`a[2]`) wasn't valid in places where it should have
been (`5[4]`, for example).

Aside from the obvious changes of adding `array` and the necessary
machinery to make this work in declarations and subroutine definitions,
this commit also separates out `indexedIdentifier` into its own entry.
This makes arbitrary levels of nested indexing on identifiers allowed in
any place that an expression is allowed.  Rules which needed to take a
simple identifier (but not an expression), such as quantum reset
operations are changed to use the new `indexedIdentifier` instead of the
old alias-based version.  This makes `reset a ++ b;` into an error,
whereas previously it was allowed by the grammar; there is currently no
basis in the specification to allow this (though it would be a simple
addition).

The separation of the `indexedIdentifier` was also necessary to make
assignment into array slices work correctly.  Before, the `designator`
rule was overloaded and used in assignments, despite it being meant for
a completely different part of the language.  An `indexedIdentifier` is
now valid as an l-value for assignments, which allows sliced,
multiple-dimensioned assignments such as

    my_array[0, 1:3] = my_array[1, 4:6];

There are some places where `indexedIdentifier` is used, though it is
possible that `indexExpression` may be more appropriate, such as the
operands of quantum gates.  For now, these are left as-is to reduce the
changeset.

### CODE REVIEW

File: examples/arrays.qasm [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+// =====================
+
+// Array arguments have a mandatory ``const`` or ``mutable`` specifier when
+// defined in a subroutine argument list.  This is because arrays are passed to
+// subroutines by reference, not by value, so modifications will propagate back
+// to the original data.  Such modifications are only allowed for ``mutable``
+// references, not ``const``.
+def copy_3_bytes(const array[uint[8], 3] in_array, mutable array[uint[8], 3] out_array) {
+    // Within this block, ``in_array`` can be read from, but not written to,
+    // whereas ``out_array`` can be both read from an written to.
```
Comment (mbhealy) [inline examples/arrays.qasm]:
```suggestion
    // whereas ``out_array`` can be both read from and written to.
```

File: source/grammar/qasm3.g4 [hunk 2]
```diff
@@ -370,20 +383,41 @@ additiveExpression
...
 
 powerExpression
+    : indexExpression
+    | powerExpression '**' indexExpression
+    ;
+
+// The general form is a comma-separated list of indexing entities.
+// `discreteSet`
+// is only valid when being used as a single index: registers can support it for
+// creating aliases, but registers cannot.
```
Comment (mbhealy) [inline source/grammar/qasm3.g4]: The part after the comma doesn't make sense, should it be: 'but arrays cannot'?

**Review by mbhealy — COMMENTED** [review]
Overall looks very good.

File: examples/arrays.qasm (hunk 1 above)
Comment (jakelishman) [reply examples/arrays.qasm]: Fixed in add15aa.

File: source/grammar/qasm3.g4 (hunk 2 above)
Comment (jakelishman) [reply source/grammar/qasm3.g4]: Ah thanks - my brain must have just given up!  Fixed in aad15aa.

**Pull request merged**